### PR TITLE
PUBDEV-8172: Support scoreContributions for imported MOJOs (in Java)

### PR DESCRIPTION
**Pull request by michalkurka**

### CODE REVIEW

File: h2o-algos/src/main/java/hex/generic/GenericModel.java [hunk 1]
```diff
@@ -167,4 +169,72 @@ MojoModel get() {
...
+                throw new RuntimeException(e);
+            }
+        }
+
+        private void predict(Chunk[] cs, NewChunk[] ncs) throws PredictException {
+            RowData rowData = new RowData();
+            byte[] types = _fr.types();
+            for (int i = 0; i < cs[0]._len; i++) {
+                RowDataUtils.extractChunkRow(cs, _fr._names, types, i, rowData);
+                ContributionsPrediction p = (ContributionsPrediction) _wrapper.predict(rowData);
```
[the remark this answers is not in the document]
Comment (valenad1) [reply h2o-algos/src/main/java/hex/generic/GenericModel.java:220]: Ad. https://h2oai.atlassian.net/browse/PUBDEV-8172?focusedCommentId=58314

Currently we can call _wrapper.predictContributions(). I will expose new interface to provide primitive types and not FeatureContributions object.

File: h2o-core/src/main/java/water/fvec/NewChunk.java [hunk 2]
```diff
@@ -1789,4 +1789,13 @@ public int cidx() {
     return _cidx;
   }
 
+  public static void addNums(NewChunk[] ncs, float[] values) {
```
Comment (valenad1) [inline h2o-core/src/main/java/water/fvec/NewChunk.java:1792]: Great method! The Javadoc here would be nice

File: h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java [hunk 3]
```diff
@@ -942,6 +942,10 @@ public CoxPHModelPrediction predictCoxPH(RowData data) throws PredictException {
   // Transparent methods passed through to GenModel.
   //----------------------------------------------------------------------
 
+  public GenModel getModel() {
```
Comment (valenad1) [inline h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java:945]: Great! Now we can make public final GenModel m; private in following PR, I'll do it.

**Review by valenad1 — COMMENTED** [review]
Generally it looks good to me, just a few points to consider

File: h2o-r/h2o-package/R/models.R [hunk 4]
```diff
@@ -907,7 +907,7 @@ h2o.predict_contributions <- predict_contributions.H2OModel
 
 #' Retrieve the number of occurrences of each feature for given observations 
 #  on their respective paths in a tree ensemble model.
-#' Available for GBM, Random Forest and Isolation Forest models.
+#' Available for GBM, Random Forest and Isolation Forest models (and equivalent imported MOJOs)
```
[the remark this answers is not in the document]
Comment (valenad1) [reply h2o-r/h2o-package/R/models.R:910]: Or make it consistent with Python message: (only GBM, XGBoost, DRF models and equivalent
        imported MOJOs).

File: h2o-algos/src/main/java/hex/generic/GenericModel.java [hunk 5]
```diff
@@ -167,4 +169,72 @@ MojoModel get() {
...
+        adaptFrm = adaptFrm.subframe(wrapper.getModel().features());
+
+        String[] outputNames = wrapper.getContributionNames();
+        return new GenericScoreContributionsTask(wrapper)
+                .withPostMapAction(JobUpdatePostMap.forJob(job))
+                .doAll(outputNames.length, Vec.T_NUM, adaptFrm)
+                .outputFrame(destination_key, outputNames, null);
+    }
+
+    class GenericScoreContributionsTask extends MRTask<GenericScoreContributionsTask> {
```
Comment (valenad1) [inline h2o-algos/src/main/java/hex/generic/GenericModel.java:192]: This class can be private

File: h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java (hunk 3 above)
Comment (valenad1) [reply h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java:945]: valenad-PUBDEV-8172-make-model-private-field

File: h2o-core/src/main/java/water/util/RowDataUtils.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+import hex.genmodel.easy.RowData;
+import water.fvec.Chunk;
+import water.fvec.Vec;
+import water.parser.BufferedString;
+
+public class RowDataUtils {
+
+    public static void extractChunkRow(Chunk[] cs, String[] names, byte[] types, int row, RowData rowData) {
+        BufferedString str = new BufferedString();
```
[the remark this answers is not in the document]
Comment (michalkurka) [reply h2o-core/src/main/java/water/util/RowDataUtils.java:11]: I think you make a good point, however, in this case, it is the responsibility of the caller to provide a suitable RowData instance.

File: h2o-r/tests/testdir_algos/generic/runit_generic_model.r [hunk 7]
```diff
@@ -0,0 +1,31 @@
+source("/Users/mkurka/git/h2o/h2o-3/h2o-r/tests/testdir_algos/generic/generic_model_test_common.R")
```
[the remark this answers is not in the document]
Comment (michalkurka) [reply h2o-r/tests/testdir_algos/generic/runit_generic_model.r:1]: thank you!

File: h2o-test-support/src/main/java/water/fvec/TestFrameBuilder.java [hunk 8]
```diff
@@ -458,8 +466,8 @@ private void checkColumnData() {
         case Vec.T_BAD:
           final double[] data = numericData.get(colIdx);
           numRows = data.length;
-          for (int i = 0; i < data.length; i++) {
-            throwIf(!Double.isNaN(data[i]), "All elements in a bad column must be NAs.");
+          for (double datum : data) {
```
[the remark this answers is not in the document]
Comment (michalkurka) [reply h2o-test-support/src/main/java/water/fvec/TestFrameBuilder.java:469]: I think in the PR I won't touch code that is just moved from one place to another

File: h2o-r/h2o-package/R/models.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (michalkurka) [reply h2o-r/h2o-package/R/models.R:910]: thank you - the documentation was also misplaced

File: h2o-genmodel/src/main/java/hex/genmodel/easy/prediction/ContributionsPrediction.java [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package hex.genmodel.easy.prediction;
```
[the remark this answers is not in the document]
Comment (michalkurka) [reply h2o-genmodel/src/main/java/hex/genmodel/easy/prediction/ContributionsPrediction.java:1]: 2 different purposes, no? 

ContributionsPrediction is a marker interface for results of EasyWrapper

File: h2o-algos/src/main/java/hex/generic/GenericModel.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (michalkurka) [reply h2o-algos/src/main/java/hex/generic/GenericModel.java:220]: PredictContributions works on a low level API, RowData is for high level API

IMHO no change needed

**Review by valenad1 — APPROVED** [review]
Thank you!

**Comment by michalkurka** [comment]
@valenad1 thank you for the thorough review; thanks to your comments we found an issue with non-default categorical encodings

**Pull request merged**